File: src/client/acontext-py/src/acontext/resources/blocks.py
"""
Block endpoints.
"""

from collections.abc import Mapping, MutableMapping
from typing import Any

from ..client_types import RequesterProtocol


class BlocksAPI:
    def __init__(self, requester: RequesterProtocol) -> None:
        self._requester = requester

    def list(
        self,
        space_id: str,
        *,
        parent_id: str | None = None,
        block_type: str | None = None,
    ) -> Any:
        params: dict[str, Any] = {}
        if parent_id is not None:
            params["parent_id"] = parent_id
        if block_type is not None:
            params["type"] = block_type
        return self._requester.request("GET", f"/space/{space_id}/block", params=params or None)

    def create(
        self,
        space_id: str,
        *,
        block_type: str,
        parent_id: str | None = None,
        title: str | None = None,
        props: Mapping[str, Any] | MutableMapping[str, Any] | None = None,
    ) -> Any:
        if not block_type:
            raise ValueError("block_type is required")
        payload: dict[str, Any] = {"type": block_type}
        if parent_id is not None:
            payload["parent_id"] = parent_id
        if title is not None:
            payload["title"] = title
        if props is not None:
            payload["props"] = props
        return self._requester.request("POST", f"/space/{space_id}/block", json_data=payload)

    def delete(self, space_id: str, block_id: str) -> None:
        self._requester.request("DELETE", f"/space/{space_id}/block/{block_id}")

    def get_properties(self, space_id: str, block_id: str) -> Any:
        return self._requester.request("GET", f"/space/{space_id}/block/{block_id}/properties")

    def update_properties(
        self,
        space_id: str,
        block_id: str,
        *,
        title: str | None = None,
        props: Mapping[str, Any] | MutableMapping[str, Any] | None = None,
    ) -> None:
        payload: dict[str, Any] = {}
        if title is not None:
            payload["title"] = title
        if props is not None:
            payload["props"] = props
        if not payload:
            raise ValueError("title or props must be provided")
        self._requester.request("PUT", f"/space/{space_id}/block/{block_id}/properties", json_data=payload)

    def move(
        self,
        space_id: str,
        block_id: str,
        *,
        parent_id: str | None = None,
        sort: int | None = None,
    ) -> None:
        payload: dict[str, Any] = {}
        if parent_id is not None:
            payload["parent_id"] = parent_id
        if sort is not None:
            payload["sort"] = sort
        if not payload:
            raise ValueError("parent_id or sort must be provided")
        self._requester.request("PUT", f"/space/{space_id}/block/{block_id}/move", json_data=payload)

    def update_sort(self, space_id: str, block_id: str, *, sort: int) -> None:
        self._requester.request(
            "PUT",
            f"/space/{space_id}/block/{block_id}/sort",
            json_data={"sort": sort},
        )

File: src/client/acontext-ts/src/resources/blocks.ts
/**
 * Block endpoints.
 */

import { RequesterProtocol } from '../client-types';
import { Block, BlockSchema } from '../types';

export class BlocksAPI {
  constructor(private requester: RequesterProtocol) {}

  async list(
    spaceId: string,
    options?: {
      parentId?: string | null;
      blockType?: string | null;
    }
  ): Promise<Block[]> {
    const params: Record<string, string | number> = {};
    if (options?.parentId !== undefined && options?.parentId !== null) {
      params.parent_id = options.parentId;
    }
    if (options?.blockType !== undefined && options?.blockType !== null) {
      params.type = options.blockType;
    }
    const data = await this.requester.request('GET', `/space/${spaceId}/block`, {
      params: Object.keys(params).length > 0 ? params : undefined,
    });
    return Array.isArray(data) ? data.map((item) => BlockSchema.parse(item)) : [];
  }

  async create(
    spaceId: string,
    options: {
      blockType: string;
      parentId?: string | null;
      title?: string | null;
      props?: Record<string, unknown> | null;
    }
  ): Promise<Block> {
    const payload: Record<string, unknown> = { type: options.blockType };
    if (options.parentId !== undefined && options.parentId !== null) {
      payload.parent_id = options.parentId;
    }
    if (options.title !== undefined && options.title !== null) {
      payload.title = options.title;
    }
    if (options.props !== undefined && options.props !== null) {
      payload.props = options.props;
    }
    const data = await this.requester.request('POST', `/space/${spaceId}/block`, {
      jsonData: payload,
    });
    return BlockSchema.parse(data);
  }

  async delete(spaceId: string, blockId: string): Promise<void> {
    await this.requester.request('DELETE', `/space/${spaceId}/block/${blockId}`);
  }

  async getProperties(spaceId: string, blockId: string): Promise<Block> {
    const data = await this.requester.request('GET', `/space/${spaceId}/block/${blockId}/properties`);
    return BlockSchema.parse(data);
  }

  async updateProperties(
    spaceId: string,
    blockId: string,
    options: {
      title?: string | null;
      props?: Record<string, unknown> | null;
    }
  ): Promise<void> {
    const payload: Record<string, unknown> = {};
    if (options.title !== undefined && options.title !== null) {
      payload.title = options.title;
    }
    if (options.props !== undefined && options.props !== null) {
      payload.props = options.props;
    }
    if (Object.keys(payload).length === 0) {
      throw new Error('title or props must be provided');
    }
    await this.requester.request('PUT', `/space/${spaceId}/block/${blockId}/properties`, {
      jsonData: payload,
    });
  }

  async move(
    spaceId: string,
    blockId: string,
    options: {
      parentId?: string | null;
      sort?: number | null;
    }
  ): Promise<void> {
    const payload: Record<string, unknown> = {};
    if (options.parentId !== undefined && options.parentId !== null) {
      payload.parent_id = options.parentId;
    }
    if (options.sort !== undefined && options.sort !== null) {
      payload.sort = options.sort;
    }
    if (Object.keys(payload).length === 0) {
      throw new Error('parentId or sort must be provided');
    }
    await this.requester.request('PUT', `/space/${spaceId}/block/${blockId}/move`, {
      jsonData: payload,
    });
  }

  async updateSort(
    spaceId: string,
    blockId: string,
    options: {
      sort: number;
    }
  ): Promise<void> {
    await this.requester.request('PUT', `/space/${spaceId}/block/${blockId}/sort`, {
      jsonData: { sort: options.sort },
    });
  }
}

